src/main/kotlin/org/jetbrains/intellij/platform/gradle/models/IvyModule.kt

```diff
@@ -87,11 +100,117 @@ data class IvyModule(
+ */
+internal fun Path.toAbsolutePathLocalPluginIvyArtifacts(): List<IvyModule.Artifact> {
+    // For local plugins we do not use relative paths.
+    // Since base path is null, the result will be an absolute path.
+    return this.explodeIntoIvyJarsArtifactsRelativeTo(null)
+}
+
+/**
+ * @see org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesHelper.createIvyArtifactRepository
  */
-internal fun Path.toIvyArtifact() = IvyModule.Artifact(
-    name = invariantSeparatorsPathString.replaceFirst(Regex("^[a-zA-Z]:/"), "/"),
-    type = when {
-        isDirectory() -> "directory"
-        else -> extension
-    },
-)
+private fun Path.explodeIntoIvyJarsArtifactsRelativeTo(basePath: Path? = null): List<IvyModule.Artifact> {
+    // The contract is that we are working with absolute normalized paths here.
+    val absNormalizedPath = this.absolute().normalize()
+    val absNormalizedBasePath = basePath?.absolute()?.normalize()
+
+    // Never ever return an Artifact pointing to a directory from this method.
+    val jars: List<Path> = if (absNormalizedPath.isDirectory()) {
+        CollectorTransformer.collectJars(absNormalizedPath)
+    } else {
+        listOf(absNormalizedPath)
+    }
+
+    val absoluteNormalizedPaths = jars.map { it.absolute().normalize() }
+    return absoluteNormalizedPaths.map { it: Path -> it.toArtifactRelativeTo(absNormalizedBasePath) }
+}
+
+private fun Path.toArtifactRelativeTo(basePath: Path?): IvyModule.Artifact {
+    // The contract is that we are working with absolute normalized paths here.
+    val absNormalizedPath = this.absolute().normalize()
+    val absNormalizedBasePath = basePath?.absolute()?.normalize()
+
+    val extString = absNormalizedPath.extension
+    val absPathStringWithoutLeading = absNormalizedPath.containingDirPathStringRelativeTo(absNormalizedBasePath)
+    // Remove the extension, if present, because the artifact pattern adds it.
+    val fileNameWithoutExt = absNormalizedPath.fileName.toString().removeSuffixIfPresent(".$extString")
+
+    return IvyModule.Artifact(
+        type = absPathStringWithoutLeading,
+        name = fileNameWithoutExt,
+        ext = extString
+    )
+}
+
+/**
+ * Returns an absolute, normalized & invariant path string of the contenting directory.
+ * The path will not have a leading path separator or drive letter for windows.
+ *
+ * Also, if the optional basePath is given, the path will be relative to it. In this case it also does not have the
+ * leading path separator.
+ *
+ * @see org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesHelper.createIvyArtifactRepository
+ */
+private fun Path.containingDirPathStringRelativeTo(basePath: Path? = null): String {
+    // The contract is that we are working with absolute normalized paths here.
+    val absNormalizedPath = this.absolute().normalize()
+    val absNormalizedBasePath = basePath?.absolute()?.normalize()
```

Comment:
So, I do this very often. The reason is that I try to design functions here as independent black boxes, which should just work by themselves. It should be more fool-proof this way.